### PR TITLE
Feature(node) - Get wallet balance at specific date or time and socket api docs

### PR DESCRIPTION
**Pull request by justinkook**
getBalanceAtTime()

- Added getWalletBalanceAtTime()
- Added @types getWalletBalanceAtTimeParams 
- Added getBalanceAtTime() in coin.ts
- Updated query for getBalanceForAddress
- Added optional --time parameter for wallet-balance.js
- Added GET route: '/:pubKey/balance/:time'
- Added aggregate collectionMethod in mockCollection.
- Added integration tests for getBalanceAtTime() and getBalance()

Documentation:

- Moved to another pull request

Issue closes: #1735 closes: #1905 closes: #1914

### CODE REVIEW

**Comment by matiu** [comment]
Great work! It is awesome that the documentation is been updated.

Would it be possible to add tests for the `getBalance` modifications and the new `getBalanceAtTime` function?

File: packages/bitcore-node/docs/wallet-guide.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+```
+touch .profile
+```
+
+Edit the .profile file to insert:
+
+```
+alias tnbitpay='/Applications/Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt -datadir=/Users/username/blockchains/bitcoin-core/networks/mainnet/'
+
+alias tnbitpayreg='/Applications/Bitcoin-Qt.app/Contents/MacOS/Bitcoin-Qt -datadir=/Users/username/blockchains/bitcoin-core/networks/regtest/'
```
Comment (micahriggan) [inline packages/bitcore-node/docs/wallet-guide.md]: let's name the aliases <chain><network>

So bitcoinregtest, bitcointestnet, bitcoinmainnet, etc

File: packages/bitcore-node/docs/wallet-guide.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+Start the Bitcore node in the /bitcore/ project root directory
+
+```
+npm run node
+```
+
+To run RegTest Bitcoin Core Client
+
+```
+. .profile
```
Comment (micahriggan) [inline packages/bitcore-node/docs/wallet-guide.md]: . ~/.profile 

If you've just run 
```
npm run node
```

Then you've likely left the homedir

File: packages/bitcore-node/docs/wallet-guide.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
+## Set up Alias config
+
+Go to the root directory of your computer
+
+```
+cd ~
+```
+
+Create a .profile file if missing
```
Comment (micahriggan) [inline packages/bitcore-node/docs/wallet-guide.md]: For some users, .bashrc will be the correct file

File: packages/bitcore-node/docs/wallet-guide.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+## How to Generate Blocks
+
+Go to Help -> Debug Window -> console tab
+
+Input generate command in the line to create 5000 BTC
+
+```
+generate 5000
```
Comment (micahriggan) [inline packages/bitcore-node/docs/wallet-guide.md]: Add a note about how now bitcore-node should be syncing blocks

File: packages/bitcore-node/src/models/coin.ts [hunk 5]
```diff
@@ -90,13 +88,31 @@ class CoinModel extends BaseModel<ICoin> {
...
+        time: { $lte: new Date(time) }
+      },
+      $orderBy: { _id: -1 }
+    });
+    const blockHeight = block!.height
+    const combinedQuery = Object.assign({}, {
+      $or: [{ spentHeight: { $gt: blockHeight } }, { spentHeight: { $lt: SpentHeightIndicators.minimum } }],
+      mintHeight: { $lte: blockHeight }
+    }, query);
+    return this.getBalance({ query: combinedQuery });
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/coin.ts]: A good unit test would be to stub the aggregation, and then expect the query passed into to aggregate to match a certain object for each method.

**Comment by justinkook** [comment]
Tests added! Ready for review

File: packages/bitcore-node/src/models/coin.ts [hunk 6]
```diff
@@ -67,19 +68,15 @@ class CoinModel extends BaseModel<ICoin> {
 
   async getBalance(params: { query: any }): Promise<{ confirmed: number, unconfirmed: number, balance: number }> {
     let { query } = params;
-    query = Object.assign(query, {
-      spentHeight: { $lt: SpentHeightIndicators.minimum },
-      mintHeight: { $gt: SpentHeightIndicators.conflicting }
-    });
-    const result =  await this.collection
+    const result = await this.collection
       .aggregate<{ _id: string, balance: number }>([
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/coin.ts]: Type here has changed, should be fixed, which would get rid of the "as" type thing after it

File: packages/bitcore-node/src/models/coin.ts [hunk 7]
```diff
@@ -67,19 +68,15 @@ class CoinModel extends BaseModel<ICoin> {
 
   async getBalance(params: { query: any }): Promise<{ confirmed: number, unconfirmed: number, balance: number }> {
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/coin.ts]: So after this change, this is really just the same as CoinStorage.collection.aggregate

**Comment by micahriggan** [comment]
Can you pull all the docs changes into their own MR?

**Pull request merged**